contributors/design-proposals/controller-ref.md

```diff
@@ -1,101 +1,375 @@
+    Adopting Pods with the wrong names would violate StatefulSet's semantics.
+
+    Adoption is allowed when Pod names match, so it remains possible to orphan a
+    family of Pods (by deleting their StatefulSet without cascading) and then
+    create a new StatefulSet with the same name and selector to adopt them.
+
+* **CronJob**
+
+  * CronJob [does not use watches](https://github.com/kubernetes/kubernetes/blob/9ae2dfacf196ca7dbee798ee9c3e1663a5f39473/pkg/controller/cronjob/cronjob_controller.go#L20),
+    so [that section](#watches) doesn't apply.
+    Instead, all CronJobs are processed together upon every "sync".
+  * CronJob applies a `created-by` annotation to link Jobs to the CronJob that
+    created them.
+    If a ControllerRef is found, it should be used instead to determine this
+    link.
+
+## Created-by annotation
+
+Aside from the change to CronJob mentioned above, several other uses of the
+`kubernetes.io/created-by` annotation have been identified that would be better
+served by ControllerRef because it tracks who *currently* controls an object,
+not just who originally created it.
+
+As a first step, the specific uses identified in the [Implementation](#implementation)
+section will be augmented to prefer ControllerRef if one is found.
+If no ControllerRef is found, they will fall back to looking at `created-by`.
+
+# Upgrading
+
+In the absence of controllers with overlapping selectors, upgrading or
+downgrading the master to or from a version that introduces ControllerRef
+should have no user-visible effects.
+If no one is fighting, adoption should always succeed eventually, so ultimately
+only the selectors matter on either side of the transition.
+
+If there are controllers with overlapping selectors at the time of an *upgrade*:
+
+* Back-and-forth thrashing should stop after the upgrade.
+* The ownership of existing objects might change due to races during
+  [adoption](#adoption). As mentioned in the [non-goals](#non-goals) section,
+  this can include breaking up families of objects that should have stayed
+  together.
+* Controllers might create additional objects because they start to respect the
+  ["Don't share"](#behavior) rule.
+
+If there are controllers with overlapping selectors at the time of a
+*downgrade*:
+
+* Controllers may begin to fight and thrash objects.
+* The ownership of existing objects might change due to ignoring ControllerRef.
+* Controllers might delete objects because they stop respecting the
+  ["Don't share"](#behavior) rule.
+
+# Implementation
+
+Checked items had been completed at the time of the last edit of this proposal
+([2017-02-01](#history)).
+
+* [x] Add API field for `Controller` to the `OwnerReference` type.
+* [x] Add validator that prevents an object from having multiple ControllerRefs.
+* [x] Add `ControllerRefManager` types to encapsulate ControllerRef manipulation
+  logic.
+* [ ] Update all affected controllers to respect ControllerRef.
```

Comment:
To be clear, I'm not talking about regular orphaning. That's done by the GC finalizer, so the controllers don't need to worry about it.

I'm talking about when controllers look for objects they own but that no longer match its selector (due to change in labels):

https://github.com/kubernetes/kubernetes/blob/9ae2dfacf196ca7dbee798ee9c3e1663a5f39473/pkg/controller/replication/replication_controller.go#L655

Are you saying controllers should still do this even after they've been marked with a DeletionTimestamp? That seems counter to the assumption made by the GC that, once a controller observes the DeletionTimestamp, it will not fight me as I go about orphaning or deleting children.